stl/inc/syncstream

```diff
@@ -0,0 +1,355 @@
+// syncstream standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _SYNCSTREAM_
+#define _SYNCSTREAM_
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+#if !_HAS_CXX20
+#pragma message("The contents of <syncstream> are available only with C++20 or later.")
+#else // ^^^ !_HAS_CXX20 / _HAS_CXX20 vvv
+#include <memory>
+#include <ostream>
+#include <shared_mutex>
+#include <streambuf>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#pragma push_macro("emit")
+#undef new
+#undef emit
+
+_STD_BEGIN
+
+extern "C" shared_mutex* __cdecl _Get_mutex_for_instance(void* _Ptr) noexcept;
+extern "C" void __cdecl _Acquire_mutex_for_instance(void* _Ptr);
+extern "C" void __cdecl _Release_mutex_for_instance(void* _Ptr) noexcept;
+
+// CLASS TEMPLATE _Basic_osyncbuf_impl
+template <class _Elem, class _Traits>
+class _Basic_syncbuf_impl : public basic_streambuf<_Elem, _Traits> {
+public:
+    void set_emit_on_sync(const bool _Val) noexcept {
+        _Emit_on_sync = _Val;
+    }
+
+    virtual bool _Do_emit() = 0;
+
+protected:
+    using _Mysb = basic_streambuf<_Elem, _Traits>;
+
+    _Basic_syncbuf_impl() = default;
+
+    _Basic_syncbuf_impl(_Basic_syncbuf_impl&& _Right) {
+        swap(_Right);
+    }
+
+    void swap(_Basic_syncbuf_impl& _Right) {
+        _Mysb::swap(_Right);
+        _STD swap(_Emit_on_sync, _Right._Emit_on_sync);
+        _STD swap(_Sync_recorded, _Right._Sync_recorded);
+    }
+
+    bool _Emit_on_sync{false};
+    bool _Sync_recorded{false};
+};
+
+// CLASS TEMPLATE basic_syncbuf
+template <class _Elem, class _Traits, class _Alloc>
+class basic_syncbuf : public _Basic_syncbuf_impl<_Elem, _Traits> {
+public:
+    using int_type       = typename _Traits::int_type;
+    using pos_type       = typename _Traits::pos_type;
+    using off_type       = typename _Traits::off_type;
+    using allocator_type = _Alloc;
+    using streambuf_type = basic_streambuf<_Elem, _Traits>;
+
+    using _Mybase   = _Basic_syncbuf_impl<_Elem, _Traits>;
+    using pointer   = typename allocator_traits<_Alloc>::pointer;
+    using size_type = typename allocator_traits<_Alloc>::size_type;

```

Comment:
As far as I can tell, `pointer` and `size_type` aren't depicted as public member types of `basic_syncbuf`. If the implementation needs these, they should be `private`. (Although we commonly uglify internal typedefs, we widely use `pointer`, and it should be okay to do the same for `size_type` - we don't need to worry about macros, we just need to ensure that users can't detect `basic_syncbuf::pointer` etc. through SFINAE.)

Reply:
I think these names need to be uglified, in case some user-defined class inherits from this class.

```c++
using pointer = void*;
struct mybuf : std::syncbuf {
    pointer p; // name lookup should find `::pointer`, not `syncbuf::pointer`
};
```